### PR TITLE
netgoal: add pre-generated network support

### PR DESCRIPTION
**Pull request by shiqizng**
This PR is for issue [#2022]([https://github.com/algorand/go-algorand/issues/2022)
netgoal generate is extended to generate a json file with structure provided in the issue. 
netgoal build now takes an additional boolean parameter -d to indicate whether to generate db files

### CODE REVIEW

**Comment by shiqizng** [comment]
currently, it takes an hour to write 1k blocks to the file

File: netdeploy/remote/deployedNetwork.go [hunk 1]
```diff
@@ -19,20 +19,28 @@ package remote
...
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/gen"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/util"
+	"github.com/algorand/go-algorand/util/codecs"
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:33]: coding style: The go-algorand imports are grouped together last, after the system imports.

File: netdeploy/remote/deployedNetwork.go [hunk 2]
```diff
@@ -266,7 +284,7 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
 	// Store genesis files in root/GenesisData
 	genesisFolder := filepath.Join(rootDir, genesisFolderName)
 	if cfg.useExistingGenesis {
-		fmt.Println(" *** using existing genesis files ***")
+		fmt.Println(" *** using ex isting genesis files ***")
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:287]: typo

File: netdeploy/remote/deployedNetwork.go [hunk 3]
```diff
@@ -63,10 +71,11 @@ type DeployedNetworkConfig struct {
...
 	useExistingGenesis bool
-
-	GenesisData gen.GenesisData
-	Topology    topology
-	Hosts       []HostConfig
+	useBoostrappedFile bool
+	GenesisData        gen.GenesisData
+	Topology           topology
+	Hosts              []HostConfig
+	BootstrappedFile   BootstrappedFile
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:78]: maybe call it `BootstrappedNetwork`, since at this point, it's no longer a file ?

File: test/testdata/deployednettemplates/recipes/scenario1/Makefile [hunk 4]
```diff
@@ -1,12 +1,16 @@
 PARAMS=-w 100 -R 8 -N 20 -n 100 -H 10 --node-template node.json --relay-template relay.json --non-participating-node-template nonPartNode.json
+FILEPARAMS=--rounds 5000 --ntrx 1000 --naccounts 3000000 --nassets 20000 --napps 20000 --wallet-name "wallet1"
```
Comment (tsachiherman) [inline test/testdata/deployednettemplates/recipes/scenario1/Makefile:2]: for now, avoid modifying the `scenario1`, since it's being used as a reference recipe. Instead, you can create a copy of that recipe.

File: netdeploy/remote/deployedNetwork.go [hunk 5]
```diff
@@ -283,9 +301,302 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+			GenesisID:   genesisID,
+			GenesisHash: genesisHash,
+			Round:       basics.Round(round),
+			RewardsState: bookkeeping.RewardsState{
+				RewardsRate: 1,
+				RewardsPool: poolAddr,
+				FeeSink:     sinkAddr,
+			},
+			UpgradeState: bookkeeping.UpgradeState{
+				CurrentProtocol: protocol.ConsensusCurrentVersion,
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:572]: use the same protocol as the one from the previous block. This is important since we sometimes need to create a simulation of `future` protocol.

File: netdeploy/remote/deployedNetwork.go [hunk 6]
```diff
@@ -283,9 +301,302 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+
+	block := bookkeeping.Block{
+		BlockHeader: bookkeeping.BlockHeader{
+			GenesisID:   genesisID,
+			GenesisHash: genesisHash,
+			Round:       basics.Round(round),
+			RewardsState: bookkeeping.RewardsState{
+				RewardsRate: 1,
+				RewardsPool: poolAddr,
+				FeeSink:     sinkAddr,
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:569]: Use the same reward pool & sink as the previous block header.

File: netdeploy/remote/deployedNetwork.go [hunk 7]
```diff
@@ -283,9 +301,302 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+func createBlock(poolAddr basics.Address, sinkAddr basics.Address, genesisID string, genesisHash crypto.Digest, round uint64, prev bookkeeping.Block, stxn []transactions.SignedTxn) (bookkeeping.Block, error) {
+	payset := make([]transactions.SignedTxnInBlock, 0, len(stxn))
+	var txibs []transactions.SignedTxnInBlock
+	txibs = make([]transactions.SignedTxnInBlock, 0, len(stxn))
+
+	block := bookkeeping.Block{
+		BlockHeader: bookkeeping.BlockHeader{
+			GenesisID:   genesisID,
+			GenesisHash: genesisHash,
+			Round:       basics.Round(round),
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:565]: try to pass the `round` as basics.Round, which would save you from doing these conversions.

File: netdeploy/remote/deployedNetwork.go [hunk 8]
```diff
@@ -283,9 +301,302 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+
+	initState.Block = initBlock
+	initState.Accounts = accounts
+	initState.GenesisHash = crypto.Hash([]byte(genesisID))
+	return initState, nil
+}
+
+func createBlock(poolAddr basics.Address, sinkAddr basics.Address, genesisID string, genesisHash crypto.Digest, round uint64, prev bookkeeping.Block, stxn []transactions.SignedTxn) (bookkeeping.Block, error) {
+	payset := make([]transactions.SignedTxnInBlock, 0, len(stxn))
+	var txibs []transactions.SignedTxnInBlock
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:558]: this line can be omitted if you'll convert the subsequent line into `:=`

File: cmd/netgoal/generate.go [hunk 9]
```diff
@@ -61,6 +67,12 @@ func init() {
 	generateCmd.Flags().StringVarP(&nodeTemplatePath, "node-template", "", "", "json for one node")
 	generateCmd.Flags().StringVarP(&nonParticipatingNodeTemplatePath, "non-participating-node-template", "", "", "json for non participating node")
 	generateCmd.Flags().StringVarP(&relayTemplatePath, "relay-template", "", "", "json for a relay node")
+	generateCmd.Flags().StringVarP(&sourceWallet, "wallet-name", "", "", "Source wallet name")
+	generateCmd.Flags().IntVarP(&rounds, "rounds", "", 5000, "Number of rounds")
+	generateCmd.Flags().IntVarP(&roundTrxCount, "ntrx", "", 1000, "Transaction count")
+	generateCmd.Flags().IntVarP(&accountsCount, "naccounts", "", 100000, "Account count")
+	generateCmd.Flags().IntVarP(&assetsCount, "nassets", "", 100000, "Asset count")
+	generateCmd.Flags().IntVarP(&applicationCount, "napps", "", 100000, "Application Count")
```
Comment (tsachiherman) [inline cmd/netgoal/generate.go:75]: I'm pretty sure that we don't want to use these numbers as the default values. 
The rounds would clearly need to be zero ( so it would be backward compatible ).

File: cmd/netgoal/generate.go [hunk 10]
```diff
@@ -61,6 +67,12 @@ func init() {
 	generateCmd.Flags().StringVarP(&nodeTemplatePath, "node-template", "", "", "json for one node")
 	generateCmd.Flags().StringVarP(&nonParticipatingNodeTemplatePath, "non-participating-node-template", "", "", "json for non participating node")
 	generateCmd.Flags().StringVarP(&relayTemplatePath, "relay-template", "", "", "json for a relay node")
+	generateCmd.Flags().StringVarP(&sourceWallet, "wallet-name", "", "", "Source wallet name")
+	generateCmd.Flags().IntVarP(&rounds, "rounds", "", 5000, "Number of rounds")
+	generateCmd.Flags().IntVarP(&roundTrxCount, "ntrx", "", 1000, "Transaction count")
```
Comment (tsachiherman) [inline cmd/netgoal/generate.go:72]: `ntrx` -> `ntxns`

**Comment by shiqizng** [comment]
it takes 30mins to generate ~2.5k blocks and 1.5hrs to generate 5k blocks with 1k transactions per block.

**Comment by shiqizng** [comment]
Updated ledger configurations as per Tsachi's suggestions. It now takes 1 minute to run 1k rounds.

**Comment by shiqizng** [comment]
this spec takes about ~30min to complete
{
	"NumRounds":30000,
	"RoundTransactionsCount": 1000,
	"GeneratedAccountsCount": 3000000,
	"GeneratedAssetsCount": 200000,
	"GeneratedApplicationCount": 1000000,
	"SourceWalletName": "wallet1"
}

File: netdeploy/remote/deployedNetwork_test.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+	sntxs, _ = createSignedTx(src, basics.Round(1), params, &networkState)
+	require.Equal(t, 2, len(sntxs))
+	require.Equal(t, protocol.PaymentTx, networkState.txState)
+	require.Equal(t, uint64(0), networkState.nApplications)
+	require.Equal(t, initialAccounts[0], accounts[0])
+	for _, sntx := range sntxs {
+		require.Equal(t, protocol.ApplicationCallTx, sntx.Txn.Type)
+	}
+
+	//	create payment transactions for the remainder rounds
```
Comment (algorandskiy) [inline netdeploy/remote/deployedNetwork_test.go:74]: a bit more consistency in comments: this file has `//comment`, `//.    comment` , `// comment`. The latter IMO preferable.

File: netdeploy/remote/deployedNetwork_test.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+	networkState.nAssets = 10
+	networkState.nApplications = 10
+	networkState.nAccounts = 1
+	networkState.assetPerAcct = 0
+	networkState.appsPerAcct = 0
+
+	params.MaxAssetsPerAccount = 5
+	//create 1 account and try to create 6 assets for the account
+	createSignedTx(src, basics.Round(1), params, &networkState)
+	for i := 0; i < 6; i++ {
```
Comment (algorandskiy) [inline netdeploy/remote/deployedNetwork_test.go:96]: <= params.MaxAssetsPerAccount

File: netdeploy/remote/deployedNetwork_test.go [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+		createSignedTx(src, basics.Round(1), params, &networkState)
+	}
+	require.Equal(t, 5, networkState.assetPerAcct)
```
Comment (algorandskiy) [inline netdeploy/remote/deployedNetwork_test.go:99]: (t, params.MaxAssetsPerAccount, ...)

File: cmd/netgoal/generate.go [hunk 14]
```diff
@@ -61,6 +67,12 @@ func init() {
 	generateCmd.Flags().StringVarP(&nodeTemplatePath, "node-template", "", "", "json for one node")
 	generateCmd.Flags().StringVarP(&nonParticipatingNodeTemplatePath, "non-participating-node-template", "", "", "json for non participating node")
 	generateCmd.Flags().StringVarP(&relayTemplatePath, "relay-template", "", "", "json for a relay node")
+	generateCmd.Flags().StringVarP(&sourceWallet, "wallet-name", "", "", "Source wallet name")
+	generateCmd.Flags().Uint64VarP(&rounds, "rounds", "", 13, "Number of rounds")
+	generateCmd.Flags().Uint64VarP(&roundTrxCount, "ntxns", "", 17, "Transaction count")
+	generateCmd.Flags().Uint64VarP(&accountsCount, "naccounts", "", 31, "Account count")
+	generateCmd.Flags().Uint64VarP(&assetsCount, "nassets", "", 5, "Asset count")
+	generateCmd.Flags().Uint64VarP(&applicationCount, "napps", "", 7, "Application Count")
```
Comment (algorandskiy) [inline cmd/netgoal/generate.go:75]: why these default values?

File: cmd/netgoal/generate.go [hunk 15]
```diff
@@ -472,3 +491,27 @@ func saveGenesisDataToDisk(genesisData gen.GenesisData, filename string) error {
 	}
 	return err
 }
+
+func generateAccountsLoadingFileTemplate(templateFilename, sourceWallet string, rounds, roundTrxCount, accountsCount, assetsCount, applicationCount uint64) error {
```
Comment (algorandskiy) [inline cmd/netgoal/generate.go:495]: nit: `tx` or `txn` are common for `transaction` instead of `trx`

File: cmd/netgoal/generate.go [hunk 16]
```diff
@@ -472,3 +491,27 @@ func saveGenesisDataToDisk(genesisData gen.GenesisData, filename string) error {
...
+
+func saveLoadingFileDataToDisk(data remote.BootstrappedNetwork, filename string) error {
+	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err == nil {
+		defer f.Close()
+
+		enc := codecs.NewFormattedJSONEncoder(f)
+		err = enc.Encode(data)
+	}
+	return err
```
Comment (algorandskiy) [inline cmd/netgoal/generate.go:516]: it is common to have a reversed check:

```suggestion
	if err != nil {
		return err
	}
	defer f.Close()
	enc := codecs.NewFormattedJSONEncoder(f)
	return enc.Encode(data)
```

File: netdeploy/remote/deployedNetwork.go [hunk 17]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+		return err
+	}
+
+	//create more transactions
+	prev, _ = l.Block(l.Latest())
+	for i := uint64(bootstrappedNet.round); i < fileCfgs.NumRounds; i++ {
+		bootstrappedNet.round++
+		blk, _ := createBlock(src, prev, fileCfgs.RoundTransactionsCount, &bootstrappedNet, params)
+		blk.BlockHeader.TimeStamp += int64(crypto.RandUint64() % 100 * 1000)
+		err = l.AddBlock(blk, agreement.Certificate{Round: bootstrappedNet.round})
```
Comment (algorandskiy) [inline netdeploy/remote/deployedNetwork.go:449]: I wonder if there are options for adding "pre-validated" blocks in batches in order to economize on signature validations.

Comment (shiqizng) [reply netdeploy/remote/deployedNetwork.go:449]: I don't see an option to add blocks in batches, but AddBlock() has validate set to false, so it doesn't actually do signature validation during eval().

File: cmd/netgoal/generate.go (hunk 14 above)
Comment (shiqizng) [reply cmd/netgoal/generate.go]: Wanted to create a small db file as default

Comment (shiqizng) [reply cmd/netgoal/generate.go]: Random values for creating a small db file as default

File: catchup/peerSelector.go [hunk 18]
```diff
@@ -185,7 +185,6 @@ func (ps *peerSelector) PeerDownloadDurationToRank(peer network.Peer, blockDownl
 	default: // i.e. peerRankInitialFourthPriority
 		return downloadDurationToRank(blockDownloadDuration, lowBlockDownloadThreshold, highBlockDownloadThreshold, peerRank3LowBlockTime, peerRank3HighBlockTime)
 
-
```
Comment (tsachiherman) [inline catchup/peerSelector.go:188]: please undo this change. you didn't really meant to make it, right ?

File: netdeploy/remote/bootstrappedNetwork.go [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+	NumRounds                 uint64
+	RoundTransactionsCount    uint64
+	GeneratedAccountsCount    uint64
+	GeneratedAssetsCount      uint64
+	GeneratedApplicationCount uint64
+	SourceWalletName          string
+}
+
+// LoadBootstrappedData loads a bootstrappedFile structure from a json file
+func LoadBootstrappedData(file string) (data BootstrappedNetwork, err error) {
```
Comment (tsachiherman) [inline netdeploy/remote/bootstrappedNetwork.go:35]: please add a unit test for `LoadBootstrappedData`, for both success and fail cases.

File: netdeploy/remote/bootstrappedNetwork.go [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+package remote
+
+import (
+	"encoding/json"
+	"os"
+)
+
+//BootstrappedNetwork contains the specs for generating db files
+type BootstrappedNetwork struct {
+	NumRounds                 uint64
```
Comment (tsachiherman) [inline netdeploy/remote/bootstrappedNetwork.go:26]: use tags here to explicitly specify the json key names. This is a generally good idea so that we'll know which structures are being serialized:
```
type BootstrappedNetwork struct {
 	NumRounds                 uint64 `json:"numRounds"`
	RoundTransactionsCount    uint64 `json:"roundTransactionsCount"`
        ...
 }
```

File: netdeploy/remote/deployedNetwork.go [hunk 21]
```diff
@@ -62,13 +73,48 @@ type DeployedNetworkConfig struct {
...
+const program = `#pragma version 2
+txn ApplicationID
+bz ok
+int 0
+byte "key"
+byte "value"
+app_local_put
+ok:
+int 1
+`
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:116]: This probably need also to go into the `BootstrappedNetwork` structure. But we can defer this to a subsequent task.

File: netdeploy/remote/deployedNetwork.go [hunk 22]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+		cfg.GenerateDatabaseFiles(cfg.BootstrappedNet, genesisFolder)
+	}
+
 	return
 }
 
+//GenerateDatabaseFiles generates database files according to the configurations
+func (cfg DeployedNetwork) GenerateDatabaseFiles(fileCfgs BootstrappedNetwork, genesisFolder string) error {
+
+	fmt.Printf("starts %v\n", time.Now())
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:353]: remove debugging prints, unless they add concrete value.

File: netdeploy/remote/deployedNetwork.go [hunk 23]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	accounts := make(map[basics.Address]basics.AccountData)
+
+	genesis, err := bookkeeping.LoadGenesisFromFile(genesisFolder + "/genesis.json")
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:356]: avoid concatenating folder and file names directly. Use `filepath.Join` instead.

File: netdeploy/remote/deployedNetwork.go [hunk 24]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	}
+
+	var bootstrappedNet netState
+
+	bootstrappedNet.genesisID = genesis.ID()
+	bootstrappedNet.genesisHash = crypto.Hash([]byte(genesis.ID()))
+	srcWallet := getGenesisAlloc(fileCfgs.SourceWalletName, genesis.Allocation)
+	if srcWallet.Address == "" {
+		return fmt.Errorf("error finding source wallet address")
+
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:368]: nit: remove empty line after `return`

File: netdeploy/remote/deployedNetwork.go [hunk 25]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	feeSink := getGenesisAlloc("FeeSink", genesis.Allocation)
+	if feeSink.Address == "" {
+		return fmt.Errorf("error finding fee sink address")
+
+	}
+	src, err := basics.UnmarshalChecksumAddress(srcWallet.Address)
+
+	poolAddr, err := basics.UnmarshalChecksumAddress(rewardsPool.Address)
+	if err != nil {
+		return err
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:386]: you might want to return here a more descriptive error - something that will tell you that the problem is with the reward pool address.

File: netdeploy/remote/deployedNetwork.go [hunk 26]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+		return fmt.Errorf("error finding source rewards ppol address")
+
+	}
+
+	feeSink := getGenesisAlloc("FeeSink", genesis.Allocation)
+	if feeSink.Address == "" {
+		return fmt.Errorf("error finding fee sink address")
+
+	}
+	src, err := basics.UnmarshalChecksumAddress(srcWallet.Address)
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:382]: add error handling for this one..

File: netdeploy/remote/deployedNetwork.go [hunk 27]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	}
+	sinkAddr, err := basics.UnmarshalChecksumAddress(feeSink.Address)
+	if err != nil {
+		return err
+	}
+
+	//initial state
+	bootstrappedNet.nAssets = fileCfgs.GeneratedAssetsCount
+	bootstrappedNet.nApplications = fileCfgs.GeneratedApplicationCount
+	//bootstrappedNet.nApplications = 0
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:396]: nit: delete this line.

File: netdeploy/remote/deployedNetwork.go [hunk 28]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	if err != nil {
+		return err
+	}
+	sinkAddr, err := basics.UnmarshalChecksumAddress(feeSink.Address)
+	if err != nil {
+		return err
+	}
+
+	//initial state
+	bootstrappedNet.nAssets = fileCfgs.GeneratedAssetsCount
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:394]: a common pattern is to use the following:
```
bootstrappedNet = netState{
     nAssets:fileCfgs.GeneratedAssetsCount,
     txnState:protocol.PaymentTx,
     ...
}
```
which would save you from duplicating the `bootstrappedNet` on every line.

File: catchup/peerSelector.go (hunk 18 above)
Comment (shiqizng) [reply catchup/peerSelector.go:188]: changes like these were applied after I ran make fmt. should I revert them?

File: netdeploy/remote/deployedNetwork.go [hunk 29]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	bootstrappedNet.nApplications = fileCfgs.GeneratedApplicationCount
+	//bootstrappedNet.nApplications = 0
+	bootstrappedNet.txnState = protocol.PaymentTx
+	bootstrappedNet.roundTxnCnt = fileCfgs.RoundTransactionsCount
+	bootstrappedNet.round = basics.Round(0)
+
+	params := config.Consensus[protocol.ConsensusCurrentVersion]
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:401]: use the protocol from the `genesis` data structure, or default to `protocol.ConsensusCurrentVersion` if it's not available.

File: netdeploy/remote/deployedNetwork.go [hunk 30]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	localCfg := config.GetDefaultLocal()
+	localCfg.Archival = true
+	localCfg.CatchpointTracking = -1
+	localCfg.LedgerSynchronousMode = 0
+	log := logging.NewLogger()
+	l, err := ledger.OpenLedger(log, genesisFolder+"/bootstrapped", false, initState, localCfg)
+	if err != nil {
+		return err
+	}
+	l.WaitForCommit(0)
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:434]: I'm surprised you found that to be needed. But it wouldn't be incorrect.

File: netdeploy/remote/deployedNetwork.go [hunk 31]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	initState, err := generateInitState(accounts, &bootstrappedNet)
+	if err != nil {
+		return err
+	}
+	localCfg := config.GetDefaultLocal()
+	localCfg.Archival = true
+	localCfg.CatchpointTracking = -1
+	localCfg.LedgerSynchronousMode = 0
+	log := logging.NewLogger()
+	l, err := ledger.OpenLedger(log, genesisFolder+"/bootstrapped", false, initState, localCfg)
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:430]: same connect as above : combine paths using `filepath.Join`

File: netdeploy/remote/deployedNetwork.go [hunk 32]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+
+	//create more transactions
+	prev, _ = l.Block(l.Latest())
+	for i := uint64(bootstrappedNet.round); i < fileCfgs.NumRounds; i++ {
+		bootstrappedNet.round++
+		blk, _ := createBlock(src, prev, fileCfgs.RoundTransactionsCount, &bootstrappedNet, params)
+		blk.BlockHeader.TimeStamp += int64(crypto.RandUint64() % 100 * 1000)
+		err = l.AddBlock(blk, agreement.Certificate{Round: bootstrappedNet.round})
+		if err != nil {
+			fmt.Printf("Error %s\n", err)
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:451]: use `%v` when printing out objects such as errors.

File: netdeploy/remote/deployedNetwork.go [hunk 33]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	l.Close()
+
+	localCfg.CatchpointTracking = 0
+	l, err = ledger.OpenLedger(log, genesisFolder+"/bootstrapped", false, initState, localCfg)
+	if err != nil {
+		return err
+	}
+	l.Close()
+
+	fmt.Printf("ends %v\n", time.Now())
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:467]: nit: drop debugging messages

File: netdeploy/remote/deployedNetwork.go [hunk 34]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	return nil
+}
+
+func getGenesisAlloc(name string, allocation []bookkeeping.GenesisAllocation) bookkeeping.GenesisAllocation {
+	for _, alloc := range allocation {
+		if strings.ToLower(alloc.Comment) == strings.ToLower(name) {
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:473]: move the `strings.ToLower(name)` outside of the loop, so that :
```
name = strings.ToLower(name)
...
     if strings.ToLower(alloc.Comment) == name {
```

File: netdeploy/remote/deployedNetwork.go (hunk 30 above)
Comment (shiqizng) [reply netdeploy/remote/deployedNetwork.go:434]: it's not needed. I meant to remove it.

File: netdeploy/remote/deployedNetwork.go [hunk 35]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+			UpgradeState: bookkeeping.UpgradeState{
+				CurrentProtocol: protocol.ConsensusCurrentVersion,
+			},
+			TxnCounter: 0,
+		},
+	}
+
+	initState.Block = block
+	initState.Accounts = accounts
+	initState.GenesisHash = crypto.Hash([]byte(bootstrappedNet.genesisID))
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:510]: Make sure you're consistent. The GenesisHash on the `initstate` need to match the one on the block header.
The one on the block header need to be calculated in the same way it was in `verifyGenesisHashes`.

File: netdeploy/remote/deployedNetwork.go [hunk 36]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	return initState, nil
+}
+
+func createBlock(src basics.Address, prev bookkeeping.Block, roundTxnCnt uint64, bootstrappedNet *netState, csParams config.ConsensusParams) (bookkeeping.Block, error) {
+	if bootstrappedNet.round%1000 == 0 {
+		fmt.Printf("round %v, time %v\n", bootstrappedNet.round, time.Now())
+	}
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:517]: remove debug messages.

File: netdeploy/remote/deployedNetwork.go [hunk 37]
```diff
@@ -62,13 +73,48 @@ type DeployedNetworkConfig struct {
...
+
+	accountsCreated bool
+	txnState        protocol.TxType
 
-	GenesisData gen.GenesisData
-	Topology    topology
-	Hosts       []HostConfig
+	round          basics.Round
+	accounts       []basics.Address
+	trxCnt         uint64
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:103]: `trxCnt` -> `txnCount`

File: netdeploy/remote/deployedNetwork.go [hunk 38]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	return block, nil
+}
+
+func generateAccounts(src basics.Address, roundTxnCnt uint64, prev bookkeeping.Block, l *ledger.Ledger, bootstrappedNet *netState, csParams config.ConsensusParams) error {
+
+	for !bootstrappedNet.accountsCreated {
+		//create accounts
+		bootstrappedNet.round++
+		blk, _ := createBlock(src, prev, roundTxnCnt, bootstrappedNet, csParams)
+		blk.BlockHeader.TimeStamp += int64(crypto.RandUint64() % 100 * 1000)
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:570]: Initialize the `blk.BlockHeader.TimeStamp` within `createBlock`

File: netdeploy/remote/deployedNetwork.go [hunk 39]
```diff
@@ -283,9 +338,424 @@ func (cfg DeployedNetwork) BuildNetworkFromTemplate(buildCfg BuildConfig, rootDi
...
+	return nil
+}
+
+func accountsNeeded(appsCount uint64, assetCount uint64, params config.ConsensusParams) uint64 {
+	var maxApps uint64
+	maxApps = uint64(params.MaxAppsCreated)
+	nAppAcct := appsCount / maxApps
+	if appsCount%maxApps != 0 {
+		nAppAcct++
+	}
```
Comment (tsachiherman) [inline netdeploy/remote/deployedNetwork.go:590]: wrap up this section in 
```
if maxApps > 0 {
  ...
```
in order to avoid division by zero.

File: test/testdata/deployednettemplates/recipes/scenario1/recipe.json [hunk 40]
```diff
@@ -4,4 +4,4 @@
     "ConfigFile": "../../configs/reference.json",
     "HostTemplatesFile": "../../hosttemplates/hosttemplates.json",
     "TopologyFile": "topology.json"
-}
+}
```
Comment (tsachiherman) [inline test/testdata/deployednettemplates/recipes/scenario1/recipe.json:7]: Add EOL

File: util/process_windows.go [hunk 41]
```diff
@@ -151,7 +151,7 @@ func getChildProcesses(pid int) []int {
 }
 
 func killProcess(pid int) error {
-	h, err := syscall.OpenProcess(syscall.SYNCHRONIZE | syscall.PROCESS_TERMINATE, false, uint32(pid))
+	h, err := syscall.OpenProcess(syscall.SYNCHRONIZE|syscall.PROCESS_TERMINATE, false, uint32(pid))
```
Comment (tsachiherman) [inline util/process_windows.go:154]: undo the change of this file.

File: catchup/peerSelector.go (hunk 18 above)
Comment (tsachiherman) [reply catchup/peerSelector.go:188]: yes. In general, you want to ensure that you apply your changes only on the files that you've modified.
( it's ok to make general codebase "fixes". But I'd rather these to be placed on a separate PR )

**Review by tsachiherman — DISMISSED** [review]
few cosmetic requests, but overall looks good.
If you could "Resolve Conversation" once you've addressed the various issues, it would be great.

File: test/testdata/deployednettemplates/networks/bootstrapped/okSpec.json [hunk 42]
```diff
@@ -0,0 +1,8 @@
+{
+  "NumRounds":65000,
+  "RoundTransactionsCount": 1000,
+  "GeneratedAccountsCount": 7000000,
+  "GeneratedAssetsCount": 200000,
+  "GeneratedApplicationCount": 1000000,
+  "SourceWalletName": "wallet1"
+}
```
Comment (tsachiherman) [inline test/testdata/deployednettemplates/networks/bootstrapped/okSpec.json:8]: add an EOL

File: test/testdata/deployednettemplates/networks/bootstrapped/okSpec.json [hunk 43]
```diff
@@ -0,0 +1,8 @@
+{
+  "NumRounds":65000,
```
Comment (tsachiherman) [inline test/testdata/deployednettemplates/networks/bootstrapped/okSpec.json:2]: don't forget that after making the Json tag names changes, you'll need to change the key names here to match.
